### ISSUE TITLE
Algorithm Security Categories

### ISSUE THREAD

**Issue by paulidale** [openssl/openssl#26156]
[NIST IR 8547](https://nvlpubs.nist.gov/nistpubs/ir/2024/NIST.IR.8547.ipd.pdf) section 4.1 introduces the concept of **Security Category**.

A security category is an integer ranging from 1 to 5 (inclusive) that specifies the relative strength of different algorithms.  Table 1 defines them thus:

| Security Category | Attack Type | Example |
| :---: | --- | ---: |
| 1 | Key search on a block cipher with a 128-bit key | AES-128 |
| 2 | Collision search on a 256-bit hash function | SHA-256 |
| 3 | Key search on a block cipher with a 192-bit key | AES-192 |
| 4 | Collision search on a 384-bit hash function | SHA3-384 |
| 5 | Key search on a block cipher with a 256-bit key | AES-256 |

It would be useful to add this as a gettable parameter to all algorithm implementations.

Some caveats include:
1. Digests get two different categories:
    - Collision Security Category
    - Preimage Security Category
2. Other algorithms are likely to get multiple categories based on intended usage.
3. We would likely want to add a _level 0_ indicating something too weak to be considered secure.
4. The category for an algorithm will probably change (downwards) over time.
5. NIST is defining the security category in the context of PQC which means classical algorithms might not get an assigned category and the analysis required to pigeonhole it could easily be nuanced, deep and involved.
6. We might want to consider making these parameters reals rather than integers to allow for future changes that place algorithms betwixt two category boundaries.  For example, an attack against SHA3-384 that weakens it to, say, 350 bits of collision resistance.

**Comment by WizardML7** [comment]
I can work on this!

**Comment by paulidale** [comment]
Feel free.  The changes will be quite pervasive I suspect.

**Comment by WizardML7** [comment]
Ok I see what you mean now from my first couple tries at implementing this for AES. I attempted to implement this by adding a SECURITY_CATEGORY param inside of the ciphercommon.c file under the cipher_known_gettable_params struct while also doing some associated changes in the ciphercommon.h file. This seemed simple enough at first but then it became apparent that doing it from the generic cipher functions means that I need to somehow get the name of the current algorithm in order to set the SECURITY_CATEGORY param that I just created. 

So I'm a bit stuck and if you have any input on where the gettable param should be included or a way to bypass this problem or any future problems you can foresee coming up from what is most likely a bit of a naive approach. Thanks!

**Comment by paulidale** [comment]
That's the right place for the parameter.

You'll need to update the various macros that instantiate the ciphers to include this parameter.  It's possible some can be done generically.  E.g. a key length of 128 bits will generally be category 1, 192 bits 3 and 256 bits 5, however there will be exceptions.

You chose what is probably the most difficult family of algorithms to start on.  I think that digests will be the simplest.

**Comment by WizardML7** [comment]
Hello,

I have a current implementation for the `ciphercommon.c` and `ciphercommon.h` files, as well as associated changes in `evp_lib.c`, `evp.h`, and `paramnames.pm`. I want to make sure that I am on what you think is the right track with these changes before applying a similar method to the digest functions.

Here is my test file to verify the actions taking place with the security parameter being set:

```c
#include <stdio.h>
#include <openssl/evp.h>
#include <openssl/params.h>
#include <openssl/core_names.h>

int main(void) {
    EVP_CIPHER *cipher;
    EVP_CIPHER_CTX *ctx;
    OSSL_PARAM params[2], *p;
    int sec_category = 0;
    int key_len = 0;

    OpenSSL_add_all_algorithms();

    cipher = EVP_CIPHER_fetch(NULL, "AES-128-CBC", NULL);
    if (!cipher) {
        fprintf(stderr, "Failed to fetch cipher\n");
        return 1;
    }

    ctx = EVP_CIPHER_CTX_new();
    EVP_CipherInit_ex(ctx, cipher, NULL, NULL, NULL, 1);

    p = params;
    *p++ = OSSL_PARAM_construct_int(OSSL_CIPHER_PARAM_SECURITY_CATEGORY, &sec_category);
    *p = OSSL_PARAM_construct_end();

    // *p++ = OSSL_PARAM_construct_int(OSSL_CIPHER_PARAM_KEYLEN, &key_len);
    // *p = OSSL_PARAM_construct_end();

    if (!EVP_CIPHER_CTX_get_params(ctx, params)) {
        fprintf(stderr, "Failed to retrieve security category\n");
        return 1;
    }

    printf("Security Category: %d\n", sec_category);

    EVP_CIPHER_free(cipher);
    EVP_CIPHER_CTX_free(ctx);
    return 0;
}
```

Additionally, here is a code diff for the changes made in providers/implementations/ciphers/ciphercommon.c:

providers/implementations/ciphers/ciphercommon.c
Diff:
```c
Original file line number	Diff line number	Diff line change
@@ -33,6 +33,7 @@ static const OSSL_PARAM cipher_known_gettable_params[] = {
    OSSL_PARAM_int(OSSL_CIPHER_PARAM_CTS, NULL),
    OSSL_PARAM_int(OSSL_CIPHER_PARAM_TLS1_MULTIBLOCK, NULL),
    OSSL_PARAM_int(OSSL_CIPHER_PARAM_HAS_RAND_KEY, NULL),
+   OSSL_PARAM_int(OSSL_CIPHER_PARAM_SECURITY_CATEGORY, 0),
    OSSL_PARAM_END
};
const OSSL_PARAM *ossl_cipher_generic_gettable_params(ossl_unused void *provctx)
@@ -96,6 +97,25 @@ int ossl_cipher_generic_get_params(OSSL_PARAM params[], unsigned int md,
        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
        return 0;
    }
+   p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_SECURITY_CATEGORY);
+   if (p != NULL) {
+       int sec_category = 0;
+
+       if (kbits == 128) {
+           sec_category = 1;
+       } else if (kbits == 192) {
+           sec_category = 3;
+       } else if (kbits == 256) {
+           sec_category = 5;
+       }
+       OSSL_PARAM_set_int(p, sec_category);
+
+       if (!OSSL_PARAM_set_int(p, sec_category)) {
+           ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+           return 0;
+       }
+   }
+
    return 1;
}

@@ -641,6 +661,11 @@ int ossl_cipher_generic_get_ctx_params(void *vctx, OSSL_PARAM params[])
        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
        return 0;
    }
+   p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_SECURITY_CATEGORY);
+   if (p != NULL && !OSSL_PARAM_set_uint(p, ctx->security_category)) {
+       ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+       return 0;
+   }
    return 1;
}

@@ -733,4 +758,15 @@ void ossl_cipher_generic_initkey(void *vctx, size_t kbits, size_t blkbits,
    ctx->blocksize = blkbits / 8;
    if (provctx != NULL)
        ctx->libctx = PROV_LIBCTX_OF(provctx); /* used for rand */

+   if (kbits == 128)
+       ctx->security_category = 1;
+   else if (kbits == 192)
+   {
+       ctx->security_category = 3;
+   }
+   else if (kbits == 256)
+   {
+       ctx->security_category = 5;
+   }
}
```
Thank you and let me know what you think!

**Comment by paulidale** [comment]
There are a few small things I'd suggest changing but the gist of this looks good.
For example, I'd suggest inequality tests against `kbits` rather than equality to future proof things:
```c
if (kbits >= 256)
     sec_category = 5;
else if (kbits >= 192)
    sec_category = 3;
else if kbits >= 128)
    sec_category = 1;
```

There will possibly be some exceptions.  For example AES XTS mode uses two keys which we push through concatenated (I don't remember offhand if the key length gets doubled or not).  GCM mode ciphers (& others) use different macros that mightn't boil down to this call.

We'll also need documentation and unit tests.

**Comment by paulidale** [comment]
I'm not going to put a hold here for a formal decision but I'd like to solicit other opinions.  @openssl/otc are we happy with this general approach to security category implementation?

**Comment by slontis** [comment]
Seems ok to me, (at least until NIST changes their mind and invents something else).
Every single algorithm would need to be tested. The Default provider contains the list of algorithms.

A question for later is how this relates to TLS.

**Comment by slontis** [comment]
There may also be some algorithms with platform specific methods (such as s390) that you would need to consider also.
This value should probably also be cached like we do with other fields.

**Comment by t-j-h** [comment]
I would stick to the plain NIST integer values. The concept is meant to be something very simple for the end user to be able to use to decide between algorithms. It is meant to be a bucket with a slightly fuzzy definition but simple enough for normal usage. 
I think making it a real or anything other than the 1-5 (or undefined) moves us to basically creating something different to the NIST intent. 

Some other calculation that holds a more nuanced or detailed value could make sense - but I think we need the simple clear NIST 1-5 values here.

**Comment by paulidale** [comment]
@t-j-h, I agree mostly.  We should stay integer with values 1 through 5 defined as NIST does.  We should also have a zero value indicating weaker than category 1 -- we have algorithms in this category already.  As Tim noted, an _undefined_ (or better _unclassified_) value is also appealing.

A real makes little sense here.  The scale isn't remotely a _nice_ mathematical function.  1, 3 and 5 are exponentially related as are 2 and 4 but the relation between 2 and 4 and 1, 3 and 5 is hard to quantify.

**Comment by t8m** [comment]
This was implemented long time ago.
